File: go/agentapi/tool_validation.go
package agentapi

import (
	"fmt"
	"strings"
)

func validateUniqueToolNames(tools []Tool) error {
	seen := make(map[string]struct{}, len(tools))
	for _, tool := range tools {
		name := strings.TrimSpace(tool.Name)
		if name == "" {
			return fmt.Errorf("tools[].name is required")
		}
		if _, ok := seen[name]; ok {
			return fmt.Errorf("duplicate tools[].name: %s", name)
		}
		seen[name] = struct{}{}
	}
	return nil
}

File: javascript/src/tool-validation.ts
import type { Tool } from "./types/tools.js";

export function validateUniqueToolNames(tools: readonly Tool[] | undefined): void {
  if (!tools) return;
  const seen = new Set<string>();
  for (const tool of tools) {
    const name = tool.name?.trim();
    if (!name) {
      throw new Error("tools[].name is required");
    }
    if (seen.has(name)) {
      throw new Error(`duplicate tools[].name: ${name}`);
    }
    seen.add(name);
  }
}
